From ddf2a2657774b33c00a9fa6363ac0aff9b457242 Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Mon, 2 Sep 2024 15:41:50 +0200
Subject: [PATCH 1177/1695] drivers: stm32_gpio: register to firewall framework

Register secure aware STM32 GPIO banks to the firewall framework
to allow GPIO and pinctrl consumer devices to load alternate
configurations for pins.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 133f4e3e41686cb5da5a864485904cec4103bb82 drivers: stm32_gpio: register to firewall framework]
Change-Id: I9b31bf5eb539b755eb194e616d673910d2d53990
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/401629
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/drivers/stm32_gpio.c | 106 ++++++++++++++++++++++++++++++++++++++
 1 file changed, 106 insertions(+)

diff --git a/core/drivers/stm32_gpio.c b/core/drivers/stm32_gpio.c
index 6784a85f7..e8995a27a 100644
--- a/core/drivers/stm32_gpio.c
+++ b/core/drivers/stm32_gpio.c
@@ -9,10 +9,12 @@
 #include <compiler.h>
 #include <drivers/clk.h>
 #include <drivers/clk_dt.h>
+#include <drivers/firewall.h>
 #include <drivers/gpio.h>
 #include <drivers/pinctrl.h>
 #include <drivers/stm32_gpio.h>
 #include <drivers/stm32_rif.h>
+#include <dt-bindings/gpio/stm32mp_gpio.h>
 #include <io.h>
 #include <kernel/dt.h>
 #include <kernel/boot.h>
@@ -868,6 +870,71 @@ out:
 	return TEE_SUCCESS;
 }
 
+/* Forward reference to stm32_gpio_set_conf_sec() defined below */
+static void stm32_gpio_set_conf_sec(struct stm32_gpio_bank *bank);
+
+static TEE_Result stm32_gpio_fw_configure(struct firewall_query *firewall)
+{
+	struct stm32_gpio_bank *bank = firewall->ctrl->priv;
+	uint32_t gpios_mask = 0;
+	bool secure = true;
+	unsigned int gpio_pos = 0;
+
+	assert(bank->sec_support);
+
+	if (firewall->arg_count != 1)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (bank->rif_cfg) {
+#ifdef CFG_STM32_RIF
+		gpio_pos = (firewall->args[0] & RIF_PER_ID_MASK) >>
+			   RIF_PER_ID_SHIFT;
+#else
+		panic("RIF support expected");
+#endif
+		gpios_mask = BIT(gpio_pos);
+
+		/* We're about to change a specific GPIO config */
+		bank->rif_cfg->access_mask[0] |= gpios_mask;
+
+		/*
+		 * Update bank RIF config with firewall configuration data
+		 * and apply it.
+		 */
+		stm32_rif_parse_cfg(firewall->args[0], bank->rif_cfg,
+				    GPIO_MAX_CID_SUPPORTED, bank->ngpios);
+		return apply_rif_config(bank, gpios_mask);
+	}
+
+	/*
+	 * Non RIF GPIO banks use TZPROT() bit mask (bits 0 to 15)
+	 * and GPIO_STM32_NSEC bit flag in the DT bindings.
+	 */
+	gpios_mask = firewall->args[0] & GENMASK_32(15, 0);
+
+	secure = !(firewall->args[0] & GPIO_STM32_NSEC);
+
+	if (gpios_mask & ~GENMASK_32(bank->ngpios, 0)) {
+		EMSG("Invalid bitmask %#"PRIx32" for GPIO bank %c",
+		     gpios_mask, 'A' + bank->bank_id);
+		return TEE_ERROR_GENERIC;
+	}
+
+	/* Update bank secure register configuration data and apply it */
+	if (secure)
+		bank->seccfgr |= gpios_mask;
+	else
+		bank->seccfgr &= ~gpios_mask;
+
+	stm32_gpio_set_conf_sec(bank);
+
+	return TEE_SUCCESS;
+}
+
+static const struct firewall_controller_ops stm32_gpio_firewall_ops = {
+	.set_conf = stm32_gpio_fw_configure,
+};
+
 static void stm32_gpio_save_rif_config(struct stm32_gpio_bank *bank)
 {
 	size_t i = 0;
@@ -1017,6 +1084,41 @@ static TEE_Result dt_stm32_gpio_bank(const void *fdt, int node,
 	return TEE_SUCCESS;
 }
 
+static TEE_Result stm32_gpio_firewall_register(const void *fdt, int node,
+					       struct stm32_gpio_bank *bank)
+{
+	struct firewall_controller *controller = NULL;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	char bank_name[] = "gpio-bank-X";
+	char *name = NULL;
+
+	if (!IS_ENABLED(CFG_DRIVERS_FIREWALL) ||
+	    !bank->sec_support)
+		return TEE_SUCCESS;
+
+	controller = calloc(1, sizeof(*controller));
+	if (!controller)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	bank_name[sizeof(bank_name) - 2] = 'A' + bank->bank_id;
+	name = strdup(bank_name);
+
+	controller->name = name;
+	controller->priv = bank;
+	controller->ops = &stm32_gpio_firewall_ops;
+
+	if (!controller->name)
+		EMSG("Warning: out of memory to store bank name");
+
+	res = firewall_dt_controller_register(fdt, node, controller);
+	if (res) {
+		free(name);
+		free(controller);
+	}
+
+	return res;
+}
+
 /* Parse a pinctrl node to register the GPIO banks it describes */
 static TEE_Result dt_stm32_gpio_pinctrl(const void *fdt, int node,
 					const void *compat_data)
@@ -1061,6 +1163,10 @@ static TEE_Result dt_stm32_gpio_pinctrl(const void *fdt, int node,
 			if (res)
 				panic();
 
+			res = stm32_gpio_firewall_register(fdt, b_node, bank);
+			if (res)
+				panic();
+
 			STAILQ_INSERT_TAIL(&bank_list, bank, link);
 		} else {
 			if (len != -FDT_ERR_NOTFOUND)
-- 
2.47.2

